virtualizarr/readers/dmrpp.py

```diff
@@ -0,0 +1,647 @@
+                group_name = os.path.dirname(fullname_tag.text).removeprefix("/")
+                groups_roots[group_name].append(var_tag)
+                dim_tags = var_tag.findall("dap:Dim", self._ns)
+                dims = self._parse_multi_dims(dim_tags)
+                group_dims[group_name].update(dims.keys())
+        # Dimensions
+        for dim_tag in root.iterfind("dap:Dimension", self._ns):
+            for g, d in group_dims.items():
+                if dim_tag.attrib["name"] in d:
+                    groups_roots[g].append(dim_tag)
+        # Attributes
+        container_attr_tag = root.find("dap:Attribute[@name='HDF5_GLOBAL']", self._ns)
+        if container_attr_tag is None:
+            attrs_tags = root.findall("dap:Attribute", self._ns)
+            for attr_tag in attrs_tags:
+                fullname_tag = attr_tag.find(
+                    "./dap:Attribute[@name='fullnamepath']/dap:Value", self._ns
+                )
+                if fullname_tag is not None and fullname_tag.text is not None:
+                    group_name = os.path.dirname(fullname_tag.text).removeprefix("/")
+                    # Add all attributes to the new dataset
+                    groups_roots[group_name].extend(attr_tag)
+        else:
+            groups_roots[next(iter(groups_roots))].extend(container_attr_tag)
+        return groups_roots
+
+    def _parse_dataset(self, root: ET.Element) -> xr.Dataset:
+        """
+        Parse the dataset using the root element of the DMR file.
+
+        Parameters
+        ----------
+        root : ET.Element
+            The root element of the DMR file.
+
+        Returns
+        -------
+        xr.Dataset
+        """
+        # Dimension names and sizes
+        dim_tags = root.findall("dap:Dimension", self._ns)
+        dataset_dims = self._parse_multi_dims(dim_tags)
+        # Data variables and coordinates
+        coord_names = self._find_coord_names(root)
+        # if no coord_names are found or coords don't include dims, dims are used as coords
+        if len(coord_names) == 0 or len(coord_names) < len(dataset_dims):
+            coord_names = set(dataset_dims.keys())
+        # Seperate and parse coords + data variables
+        coord_vars: dict[str, xr.Variable] = {}
+        data_vars: dict[str, xr.Variable] = {}
+        for var_tag in self._find_var_tags(root):
+            variable = self._parse_variable(var_tag, dataset_dims)
+            if var_tag.attrib["name"] in coord_names:
+                coord_vars[var_tag.attrib["name"]] = variable
+            else:
+                data_vars[var_tag.attrib["name"]] = variable
+        # Attributes
+        attrs: dict[str, str] = {}
+        for attr_tag in self.root.iterfind("dap:Attribute", self._ns):
+            attrs.update(self._parse_attribute(attr_tag))
+        return xr.Dataset(
+            data_vars=data_vars,
+            coords=xr.Coordinates(coords=coord_vars, indexes={}),
```

Comment:
Should the `indexes` variable here technically be passed through from `open_virtual_dataset`? Right now it doesn't matter, but it would in order to support #18.

Reply:
Yes you're right, I will fix that

Reply:
This and https://github.com/zarr-developers/VirtualiZarr/pull/113#discussion_r1729309485 are the only comment that I think actually need to be addressed before merging, because they affect public API. Everything else can be left until later if you prefer.

Reply:
I was trying to get this fixed however I think there is an issue which is that the indexes cannot be created since we don't have the path to the actual data file. Technically the dmrpp file does have a data path but it is only the file name and not the full path. E.g. "data.nc" instead of "s3://.../data.nc". So in that case if I add support for the indexes param it will fail when `indexes=None` when usually in virtualizarr `indexes=None` indicates auto-create indexes.

What I could do instead is to treat `indexes=None` and `indexes={}` as the same but raise a warning if `indexes=None` that the behavior is not as expected. However I can still accept manually created indexes (e.g. {"time": Index})

Reply:
You can just raise a `NotImplementedError` on `indexes={}`. But at least something that isn't silently doing an inconsistent behaviour to the other readers.

Reply:
I thought `indexes={}` would be acceptable since that is how most people indicate that indexes should not be created?

Reply:
I meant indexes=None sorry